protocol/x/subaccounts/keeper/isolated_subaccount.go

```diff
@@ -0,0 +1,144 @@
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dydxprotocol/v4-chain/protocol/lib"
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	"github.com/dydxprotocol/v4-chain/protocol/x/subaccounts/types"
+)
+
+// checkIsolatedSubaccountConstaints will validate all `updates` to the relevant subaccounts against
+// isolated subaccount constraints.
+// The `updates` have to contain `Subaccounts` with unique `SubaccountIds`.
+// The input subaccounts must be settled.
+//
+// Returns a `success` value of `true` if all updates are valid.
+// Returns a `successPerUpdates` value, which is a slice of `UpdateResult`.
+// These map to the updates and are used to indicate which of the updates
+// caused a failure, if any.
+func (k Keeper) checkIsolatedSubaccountConstraints(
+	ctx sdk.Context,
+	settledUpdates []settledUpdate,
+	perpetuals []perptypes.Perpetual,
+) (
+	success bool,
+	successPerUpdate []types.UpdateResult,
+	err error,
+) {
+	success = true
+	successPerUpdate = make([]types.UpdateResult, len(settledUpdates))
+	var idOfSettledUpdates = make(map[types.SubaccountId]struct{})
+	var perpIdToMarketType = make(map[uint32]perptypes.PerpetualMarketType)
+
+	for _, perpetual := range perpetuals {
+		perpIdToMarketType[perpetual.GetId()] = perpetual.Params.MarketType
+	}
+
+	for i, u := range settledUpdates {
+		_, exists := idOfSettledUpdates[*u.SettledSubaccount.Id]
+
+		if exists {
+			return false, nil, types.ErrNonUniqueUpdatesSubaccount
+		}
+
+		result, err := isValidIsolatedPerpetualUpdates(u, perpIdToMarketType)
+		if err != nil {
+			return false, nil, err
+		}
+		if result != types.Success {
+			success = false
+		}
+
+		successPerUpdate[i] = result
+		idOfSettledUpdates[*u.SettledSubaccount.Id] = struct{}{}
+	}
+
+	return success, successPerUpdate, nil
+}
+
+// Checks whether the perpetual updates to a settled subaccount violates constraints for isolated
+// perpetuals. This function assumes the settled subaccount is valid and does not violate the
+// the constraints.
+// The constraint being checked is:
+//   - a subaccount with a position in an isolated perpetual cannot have updates other perpetuals
+//   - a subaccount with a position in a non-isolated perpetual cannot have updates for isolated
+//     perpetuals
+//   - a subaccount with no positions cannot be updated to have positions in multiple isolated
+//     perpetuals or a combination of isolated and non-isolated perpetuals
+func isValidIsolatedPerpetualUpdates(
+	settledUpdate settledUpdate,
+	perpIdToMarketType map[uint32]perptypes.PerpetualMarketType,
+) (types.UpdateResult, error) {
+	// If there are no perpetual updates, then this update does not violate constraints for isolated
+	// markets.
+	if len(settledUpdate.PerpetualUpdates) == 0 {
+		return types.Success, nil
+	}
+
+	// Check if the updates contain an update to an isolated perpetual.
+	hasIsolatedUpdate := false
+	isolatedUpdatePerpetualId := uint32(0)
```

Comment:
Is 0 a currently valid perpetual id? If so, can we set this to a value like -1?

Reply:
perpetual ids are `uint32` so they can't be set to -1. 0 is a valid id, all unsigned integers are valid perpetual ids though.

Reply:
I also don't have strong opinons on this and would prefer an invalid id. I can use MAX_UINT32 as well, might be a better number since it's unlikely to be a valid perpetual id.